code/controllers/subsystem/weather.dm

```diff
@@ -8,24 +8,66 @@ SUBSYSTEM_DEF(weather)
-		var/datum/weather/our_event = V
-		if(our_event.aesthetic || our_event.stage != MAIN_STAGE)
+	for(var/datum/weather/weather_event in processing)
+		if(IS_WEATHER_AESTHETIC(weather_event.weather_flags) || weather_event.stage != MAIN_STAGE)
 			continue
-		for(var/mob/act_on as anything in GLOB.mob_living_list)
-			if(our_event.can_weather_act(act_on))
-				our_event.weather_act(act_on)
+
+		if(weather_event.currentpart == SSWEATHER_MOBS)
+			if(!resumed)
+				weather_event.current_mobs = GLOB.mob_living_list.Copy()
+			var/list/current_mobs_cache = weather_event.current_mobs // cache for performance
+			while(current_mobs_cache.len)
+				var/mob/living/target = current_mobs_cache[current_mobs_cache.len]
+				current_mobs_cache.len--
+				if(QDELETED(target))
+					continue
+				if(weather_event.can_weather_act_mob(target))
+					weather_event.weather_act_mob(target)
+				if(MC_TICK_CHECK)
+					return
+			resumed = FALSE
+			weather_event.next_subsystem_task()
+
+		if(weather_event.currentpart == SSWEATHER_TURFS)
+			if(!resumed)
+				weather_event.turf_iteration = floor(weather_event.weather_turfs_per_tick)
+				var/fraction_chance = fract(weather_event.weather_turfs_per_tick) * 100
+				weather_event.turf_iteration += prob(fraction_chance) ? 1 : 0
+			var/list/weather_turfs = weather_event.weather_turfs // cache for performance
+			while(weather_event.turf_iteration)
+				weather_event.turf_iteration--
+				var/turf/open/selected_turf = pick(weather_turfs)
+				weather_event.weather_act_turf(selected_turf)
+				if(MC_TICK_CHECK)
+					return
+			resumed = FALSE
+			weather_event.next_subsystem_task()
+
+		if(weather_event.currentpart == SSWEATHER_THUNDER)
+			if(!resumed)
+				weather_event.thunder_iteration = floor(weather_event.thunder_turfs_per_tick)
+				var/fraction_chance = fract(weather_event.thunder_turfs_per_tick) * 100
+				weather_event.thunder_iteration += prob(fraction_chance) ? 1 : 0
+			var/list/weather_turfs = weather_event.weather_turfs // cache for performance
+			while(weather_event.thunder_iteration)
+				weather_event.thunder_iteration--
+				var/turf/open/selected_turf = pick(weather_turfs)
+				weather_event.thunder_act_turf(selected_turf)
+				if(MC_TICK_CHECK)
+					return
+			resumed = FALSE
+			weather_event.next_subsystem_task()
 
 	// start random weather on relevant levels
 	for(var/z in eligible_zlevels)
 		var/possible_weather = eligible_zlevels[z]
-		var/datum/weather/our_event = pick_weight(possible_weather)
-		run_weather(our_event, list(text2num(z)))
+		var/datum/weather/weather_event = pick_weight(possible_weather)
+		run_weather(weather_event, list(text2num(z)))
 		eligible_zlevels -= z
-		var/randTime = rand(3000, 6000)
-		next_hit_by_zlevel["[z]"] = addtimer(CALLBACK(src, PROC_REF(make_eligible), z, possible_weather), randTime + initial(our_event.weather_duration_upper), TIMER_UNIQUE|TIMER_STOPPABLE) //Around 5-10 minutes between weathers
+		var/randTime = rand(5 MINUTES, 10 MINUTES)
```

Comment:
❤️

Reply:
Small recognitions like this are nice for any code reviews because they are grueling. Ty